automation/taskcluster/lib/tasks.py

```diff
@@ -2,143 +2,517 @@
+                'tier': 1,
+            },
+        )
+
+    def craft_push_task(
+        self, signing_task_id, apks, is_staging=True, commit=False
+    ):
+        payload = {
+            "commit": commit,
+            "google_play_track": 'nightly',
+            "upstreamArtifacts": [
+                {
+                    "paths": apks,
+                    "taskId": signing_task_id,
+                    "taskType": "signing"
+                }
+            ]
         }
 
+        return self._craft_default_task_definition(
+            worker_type='mobile-pushapk-dep-v1' if is_staging else 'mobile-pushapk-v1',
+            provisioner_id='scriptworker-prov-v1',
+            dependencies=[signing_task_id],
+            routes=[],
+            scopes=[
+                "project:mobile:fenix:releng:googleplay:product:fenix{}".format(
+                    ':dep' if is_staging else ''
+                )
+            ],
+            name="Push task",
+            description="Upload signed release builds of Fenix to Google Play",
+            payload=payload,
+            treeherder={
+                'jobKind': 'other',
+                'machine': {
+                  'platform': 'android-all',
+                },
+                'symbol': 'gp',
+                'tier': 1,
+            },
+        )
+
+
+def _craft_treeherder_platform_from_variant(variant):
+    architecture, build_type, _ = _get_architecture_and_build_type_and_product_from_variant(
+        variant
+    )
+    return 'android-{}-{}'.format(architecture, build_type)
+
+
+def _craft_treeherder_group_symbol_from_variant(variant):
+    _, __, product = _get_architecture_and_build_type_and_product_from_variant(variant)
+    return product
+
+
+def _craft_artifacts_from_variant(variant):
+    return {
+        'public/target.apk': {
+            'type': 'file',
+            'path': _craft_apk_full_path_from_variant(variant),
+            'expires': taskcluster.stringDate(taskcluster.fromNow(DEFAULT_EXPIRES_IN)),
+        }
+    }
+
+
+def _craft_apk_full_path_from_variant(variant):
+    architecture, build_type, product = _get_architecture_and_build_type_and_product_from_variant(
+        variant
+    )
+
+    short_variant = variant[:-len(build_type)]
+    postfix = '-unsigned' if build_type == 'release' else ''
+    product = '{}{}'.format(product[0].lower(), product[1:])
+
+    return '/opt/fenix/app/build/outputs/apk/{short_variant}/{build_type}/app-{architecture}-{product}-{build_type}{postfix}.apk'.format(     # noqa: E501
+        architecture=architecture,
+        build_type=build_type,
+        product=product,
+        short_variant=short_variant,
+        postfix=postfix
+    )
+
+
+_SUPPORTED_ARCHITECTURES = ('aarch64', 'arm', 'x86')
+_SUPPORTED_BUILD_TYPES = ('Debug', 'Release')
+_SUPPORTED_PRODUCTS = ('FirefoxBeta', 'FirefoxNightly', 'FirefoxRelease', 'Greenfield')
+
+
+def _get_architecture_and_build_type_and_product_from_variant(variant):
```

Comment:
Perhaps instead of having `_get_x_and_y_and_z_and_<snip>(...)` functions called throughout this file, perhaps we can parse the `variant` string into a little class at the beginning, then pass an instance of that class around, rather than the string?

Reply:
Sounds good. I'd say let's do it if we need to add a 4th element. I'd prefer to keep the output as simple as possible, in order to avoid mocking functions out, whenever we need to test them.

Reply:
IMHO, a data-y class would fit the bill, which would be sufficiently simple that you wouldn't need to mock functions to test with it.

For example, I'd imagine it would look something like:
```
class Variant:
    def __init__(self, arch, build_type, product):
        self.arch = arch
        self.build_type = build_type
        self.product = product
   
   @staticmethod
    def from_variant_string(variant):
        # the logic from `_get_architecture_and_build_type_and_product_from_variant
        return Variant(arch, build_type, product)

# Now, within tasks.py
def craft_whatever_task(variant):
    print(variant.build_type)
    # etc
```

If we were to do testing of `Variant` itself, we'd need to check the `from_variant_string` function.
However, throughout the rest of our automation logic, we can just use the `Variant` class as-is, since it's so simple that there's no object functions to mock:

```
def test_thing():
    mock_variant = Variant('mock arch', 'mock build type', 'mock product')
    result = craft_whatever_task(mock_variant)
    assert(... == result)
```

-----

Though this isn't high-impact for this specific case/PR, I just wanted to illustrate how, IMHO, such a class wouldn't obfuscate testing, while it could still simplify development for production behaviour?
What do you think?

Reply:
Yeah, that would definitely work. We can even leverage https://docs.python.org/3/library/collections.html#collections.namedtuple (py2-compatible) or data-classes https://docs.python.org/3/library/dataclasses.html (py3.7+). Let's address this in a followup